asserts/fetcher.go

```diff
@@ -84,26 +136,42 @@ func (f *fetcher) chase(ref *Ref, a Assertion) error {
 		}
 		a = retrieved
 	}
-	f.fetched[u] = fetchRetrieved
-	for _, preref := range a.Prerequisites() {
-		if err := f.Fetch(preref); err != nil {
-			return err
-		}
+	return f.fetchPrerequisitesAndSave(ref.Unique(), a)
+}
+
+// Fetch retrieves the assertion indicated by ref then its prerequisites
+// recursively, along the way saving prerequisites before dependent assertions.
+func (f *fetcher) Fetch(ref *Ref) error {
+	return f.chase(ref, nil)
+}
+
+func (f *fetcher) wasSeqFetched(seq *AtSequence) (bool, error) {
+	u := seq.Unique()
+	switch f.fetched[u] {
+	case fetchSaved:
+		return true, nil // nothing to do
+	case fetchRetrieved:
+		return false, fmt.Errorf("circular assertions are not expected: %s", seq)
```

Comment:
I wonder if this is tested? And if not if it's worth adding a check?

Reply:
I added tests for circular reference detection for both the original fetcher and the new! Thanks for pointing this out!